### PR TITLE
feat(dataset): dataset creation with s3 storage backend

### LINKED ISSUE #2970: Create dataset with a S3 storage backend
As a Renku user I want to be able to create dataset with a S3 storage backend so that I can load and store data on a S3
bucket.
```
renku dataset create my-dataset --storage s3://server/path/to/bucket
```

## Acceptance Criteria
- Add a `--storage` flag to `dataset create` which accepts a URI
- S3 URI is stored in dataset's metadata
- Dataset's data directory is added to .gitignore
- Renku asks for S3 credentials (if not available) and asks the user whether to store them or not. It should store them in `~/.renku/renku.ini`
- Renku must check if S3 bucket exists and is accessible. The operation must fail otherwise.

## Notes
- When implementing this consider the use-case of accessing files from an external storage but outside of a Renku dataset.

### LINKED ISSUE #2972: Show information about external storage
As a Renku user I want to be able to use `dataset show/ls-files` commands to know external storage sources for datasets and files.

## Acceptance Criteria
- Add a `Data Sources` section to `dataset show` that lists all used external sources in a dataset including its type, URI and location on the filesystem
- Add a `data source` columns to `dataset ls-files` to show the external source for each file

### PR DESCRIPTION
**Pull request by mohammad-sdsc**
# Description

Allows creating datasets with S3 storage backends. Currently, only supports `s3://<bucket-name>/<path>` URI format. Uses `RClone` to interact with S3.

### NOTE:
`RClone` cannot check if a directory exists for storage backends that don't support empty directories like S3. So, it's not possible to fail if the directory doesn't exists as was required in the story.

Fixes #2970 
Fixes #2972 

### TODO
Test with private S3 buckets and valid S3 credentials.

### CODE REVIEW

File: renku/infrastructure/storage/base.py
Comment (Panaetius) [inline renku/infrastructure/storage/base.py:83]:
```suggestion
    args =[]
    for k, v in kwargs.items():
        args.extend(transform_kwarg(k, v))
    return args
```

not that it matters at all, feel free to ignore, i just find it slightly more readable than the list comprehension in this case.

File: renku/core/dataset/providers/api.py
Comment (Panaetius) [inline renku/core/dataset/providers/api.py:242]: Together with the other comment on `None` instead of `""`. We do have `renku.core.util.util.NO_VALUE` (And `NoValueType`) for this purpose, which allows showing intent more clearly and also reflects this in the type signature.

File: renku/core/dataset/providers/api.py
Comment (Panaetius) [inline renku/core/dataset/providers/api.py:252]: Should this just return `List[str]` as per the docstring?

**Review by Panaetius — CHANGES_REQUESTED** [review]
Looks really nice and should be easy to extend, thank you. Only some minor comments.

File: renku/infrastructure/storage/base.py
Comment (olevski) [inline renku/infrastructure/storage/base.py:29]: This currently has a lot of rclone-specific things. Should it then be called `RcloneBaseStorage` or something like that. Or if this is supposed to be more general then the rclone-specific stuff should then move someplace else?

File: renku/core/dataset/providers/s3.py
Comment (mohammad-sdsc) [reply renku/core/dataset/providers/s3.py:3]: I've changed all of them to `2017-2022`

File: renku/core/dataset/providers/api.py
Comment (mohammad-sdsc) [reply renku/core/dataset/providers/api.py:242]: `""` is a valid value. So, I've changed `None` to `NO_VALUE`.

File: renku/core/dataset/providers/api.py
Comment (mohammad-sdsc) [reply renku/core/dataset/providers/api.py:252]: This won't change and makes sense to be a tuple. I've updated the docstring.

File: renku/domain_model/dataset.py
Comment (olevski) [inline renku/domain_model/dataset.py:321]: Do we really need to have `storage` as both a class and instance variable? Isn't it simpler to just have the instance variable?

But maybe I am missing something. This also seems to happen with a few other class/instance variables on dataset.

**Comment by olevski** [comment]
Also (if I am not mistaken) the credentials can only be added for a single s3 bucket. So I believe that the way this is setup is that I can only add buckets (with credentials) from a single region and provider.

If I am wrong about this lets touch base over slack @mohammad-sdsc because I probably misunderstood of how the credential provider is supposed to be used.

But my impression is that `s3://` as a prefix is only tied to one set of credentials. And we are allowing only `s3://` to be used as a prefix for s3 buckets.

File: renku/domain_model/dataset.py
Comment (mohammad-sdsc) [reply renku/domain_model/dataset.py:321]: We need the class variable for backward compatibility: When we load an object from an older project that doesn't have the `storage` property, the class variable makes sure that this property is set to `None`. The library that we use to store Renku metadata (i.e. ZODB/zope/persistent) doesn't call the `__init__` method when deserializing an object (it creates an empty object and calls `__setstate__` upon the first access to the object). So, we need the class variable for every new attribute that we add to our models.

**Comment by mohammad-sdsc** [comment]
> Also (if I am not mistaken) the credentials can only be added for a single s3 bucket. So I believe that the way this is setup is that I can only add buckets (with credentials) from a single region and provider.
> 
> If I am wrong about this lets touch base over slack @mohammad-sdsc because I probably misunderstood of how the credential provider is supposed to be used.
> 
> But my impression is that `s3://` as a prefix is only tied to one set of credentials. And we are allowing only `s3://` to be used as a prefix for s3 buckets.

That's true, atm we only can have one set of credentials for S3. We can have multiple credentials by overriding `ProviderCredentials::get_credentials_section_name` for S3 (the comment there reads: `NOTE: This methods should be overridden by subclasses to allow multiple credentials per providers if needed.`). I added bucket name as the part of key before but then we needed to enter the credentials for each bucket so I removed it. I can easily put it back or implement any other mechanism.

**Review by olevski — APPROVED** [review]

**Comment by olevski** [comment]
This LGTM but if you want to wait also for Ralf's approval that is def fine by me.

**Review by Panaetius — APPROVED** [review]

**Pull request merged**
